### PR TITLE
[lodash] Fix _.partial() with 3 arity func and full args

### PR DESCRIPTION
**Pull request by yuu-no**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://lodash.com/docs/4.17.15#partial
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

https://www.typescriptlang.org/play/?ssl=13&ssc=9&pln=12&pc=1#code/JYWwDg9gTgLgBAfQDRwN5wGIFcB2BjGYCHFAGQgBEBDAZwAsBlGKwvOAXzgDMoIQ4A5ABsIAE1p0BAWABQsvMRrwuARjgBeOAAoqALjg4sIAEYBTKAEoNAPjhUA3PMXKATBu16DRs1BTH9SlDAOADmVuq2VHAA1HDGjjIKOErcAMzuOvqGJuZ+ATBBoSh4+sYQEEKmVDjhkTFx9XgJSSlcACwZntk+eXCBwSHFpeWV1SiiWd7mtXb1xo31oglOyfBgapoIAHRgVLDAVEIAPADaALoo3bleOVDWWqooKgAMFgkA9O9w33AAegD8KxSYDcmx2e0IhweLiezxQAioxjwAjesk+Pz+gMSzjgYHSYN2+yhXFSsPhiORKAKWFMqJk6J+AKBaw6BIhByEDzaZMEFIEVKgNNhdIZ3yZMiAA

`p3` must be `() => string`. But below.

```ts
const p3 = _.partial(f3, 10, 'abc', true);
//    ^? const p3: _.Function1<unknown, string>
```

`<T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, arg2: T2, arg3: T3): Function1<T4, R>;` brake `        <TS extends any[], T1, T2, T3, R>(func: (t1: T1, t2: T2, t3: T3, ...ts: TS) => R, t1: T1, t2: T2, t3: T3): (...ts: TS) => R;
` by overload

### CODE REVIEW

File: types/lodash/common/function.d.ts [hunk 1]
```diff
@@ -655,7 +655,6 @@ declare module "../index" {
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, plc2: __, arg3: T3): Function3<T1, T2, T4, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, plc2: __, arg3: T3): Function2<T2, T4, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, arg2: T2, arg3: T3): Function2<T1, T4, R>;
-        <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, arg2: T2, arg3: T3): Function1<T4, R>;
```
Comment (aj-r) [inline types/lodash/common/function.d.ts:658]: This seems to be working as intended. I don't think it should be removed.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -7295,7 +7298,16 @@ _.templateSettings; // $ExpectType TemplateSettings
     _.partial(func2, _.partial.placeholder, "foo"); // $ExpectType Function1<number, number>
     _.partial(func2, 42, "foo"); // $ExpectType () => number
     // with arity 3 function
-    _.partial(func3, 42,     _, true);
+    _.partial(func3) // $ExpectType (arg1: number, arg2: string, arg3: boolean) => number
+    _.partial(func3, 42) // $ExpectType (arg2: string, arg3: boolean) => number
+    _.partial(func3, 42,     _, true) // $ExpectType Function1<string, number>;
+    _.partial(func3, 42, "foo", true) // $ExpectType () => number
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:7304]: This test essentially passes even without your changes (although the expected type would be `Function0<number>` which is basicalyl the same).

**Review by aj-r — CHANGES_REQUESTED** [review]
I don't think this change is necessary. When I test your example, it works as expected: https://stackblitz.com/edit/typescript-lodash-playground-1tpr1g?file=index.ts. In that code, `p3` has type `Function0<string>` as expected.

Perhaps the issue is that your `f3` has multiple overloads with different numbers of arguments? If so, you probably want to avoid `partial` since its behavior will be ambiguous.

Thanks for adding the missing tests though - optionally you could just revert the change to `function.d.ts` and sumbit the new tests.

File: types/lodash/common/function.d.ts (hunk 1 above)
Comment (YuseiUeno) [reply types/lodash/common/function.d.ts:658]: Really?
It looks not working on `pnpm test`.

```diff
❯ git --no-pager diff --cached
diff --git a/types/lodash/common/function.d.ts b/types/lodash/common/function.d.ts
index 46ac3b3746..7628278464 100644
--- a/types/lodash/common/function.d.ts
+++ b/types/lodash/common/function.d.ts
@@ -655,6 +655,7 @@ declare module "../index" {
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, plc2: __, arg3: T3): Function3<T1, T2, T4, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, plc2: __, arg3: T3): Function2<T2, T4, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, arg2: T2, arg3: T3): Function2<T1, T4, R>;
+        <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, arg2: T2, arg3: T3): Function1<T4, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, plc2: __, plc3: __, arg4: T4): Function3<T1, T2, T3, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, plc2: __, plc3: __, arg4: T4): Function2<T2, T3, R>;
         <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, plc1: __, arg2: T2, plc3: __, arg4: T4): Function2<T1, T3, R>;
```

```
❯ pnpm -w run test lodash

> definitely-typed@0.0.3 test /Users/yusei/Workspace/repos/github.com/DefinitelyTyped/DefinitelyTyped
> node --enable-source-maps node_modules/@definitelytyped/dtslint/ types "lodash"

dtslint@0.2.20
Error:
/Users/yusei/Workspace/repos/github.com/DefinitelyTyped/DefinitelyTyped/types/lodash/lodash-tests.ts
  7304:5  error  TypeScript@4.7, 4.8, 4.9, 5.0, 5.1, 5.2, 5.3, 5.4, 5.5 expected type to be:
  () => number
got:
  Function1<unknown, number>  @definitelytyped/expect

✖ 1 problem (1 error, 0 warnings)

    at combineErrorsAndWarnings (/Users/yusei/Workspace/repos/github.com/DefinitelyTyped/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.2.20_typescript@5.5.0-dev.20240509/node_modules/@definitelytyped/dtslint/src/index.ts:259:26)
    at runTests (/Users/yusei/Workspace/repos/github.com/DefinitelyTyped/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.2.20_typescript@5.5.0-dev.20240509/node_modules/@definitelytyped/dtslint/src/index.ts:250:18)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at main (/Users/yusei/Workspace/repos/github.com/DefinitelyTyped/DefinitelyTyped/node_modules/.pnpm/@definitelytyped+dtslint@0.2.20_typescript@5.5.0-dev.20240509/node_modules/@definitelytyped/dtslint/src/index.ts:103:22)
 ELIFECYCLE  Test failed. See above for more details.
```

Comment (YuseiUeno) [reply types/lodash/common/function.d.ts:658]: And this pattern is followed here so I delete it.
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d293cb430e2778a9f2446ddd07e7f6eb8f967cba/types/lodash/common/function.d.ts#L669

Comment (YuseiUeno) [reply types/lodash/common/function.d.ts:658]: There is not working.
https://stackblitz.com/edit/fjjtuv?file=index.ts

May be your stackblitz dependency is too old.

**Comment by YuseiUeno** [comment]
> <T1, T2, T3, T4, R>(func: Function4<T1, T2, T3, T4, R>, arg1: T1, arg2: T2, arg3: T3): Function1<T4, R>; brake        <TS extends any[], T1, T2, T3, R>(func: (t1: T1, t2: T2, t3: T3, ...ts: TS) => R, t1: T1, t2: T2, t3: T3): (...ts: TS) => R; by overload

This is Function Overloads Ordering Problem.
https://github.com/microsoft/TypeScript-Handbook/blob/master/pages/declaration%20files/Do's%20and%20Don'ts.md#ordering

File: types/lodash/common/function.d.ts (hunk 1 above)
Comment (aj-r) [reply types/lodash/common/function.d.ts:658]: Ah, yeah I missed the overload below which does this. Thanks!

**Review by aj-r — APPROVED** [review]

**Review by sandersn — APPROVED** [review]

**Comment by YuseiUeno** [comment]
Ready to merge

**Pull request merged**